Store meta-data in observations as JSON

tomaszmrugalski opened the pull request:
This is definitely a WIP, but the overall goal is to be able to store meta-data with the observations. By meta-data, I mean many aspects. Some examples are:

- SDR used
- LNA, filter being used
- what kind fo antenna is used
- what kind of recipe is used
- etc.

tomaszmrugalski commented:
That's how the observations list looks like now. Two extra columns were added: frequency and protocol. Many more are stored in the config field as JSON.

<img width="1157" alt="Screenshot 2021-11-27 at 00 59 33" src="https://user-images.githubusercontent.com/663576/143661764-2079c586-bd41-4627-9ac0-2cf29c27e519.png">

All available parameters are printed in the station and observation views.

tomaszmrugalski commented:
This is probably enough for an early review. In particular, I'm interested if this approach is OK or not. If it is, I'll go ahead and modify the submit_obs code to provide parameters for new observations.

The upgrade script retroactively sets the configuration parameters (frequency, hardware used etc).

fivitti commented on server/db/svarog-11.psql:
```diff
@@ -0,0 +1,34 @@
+update observations set config = config || '{"antenna": "WiMo TA-1", "antenna-type": "crossed dipole" }' WHERE station_id = 1;
+
+-- Add frequencies. We haven't done any Doppler magic, so everything was received on base frequencies.
+update observations set config = config || '{"freq": 137.62 }' WHERE sat_id = 25338 and station_id = 1; -- NOAA 15
```
It should be "frequency". If we will promote the abbreviation then one user will use "freq" and another "f".

fivitti commented on server/db/svarog-11.psql:
```diff
@@ -0,0 +1,34 @@
+-- the PR for adding generic RX recipe (PR#50) was merged. The merge was on 2021-06-09, so let's assume
+-- that's the date of AirSpy upgrade.
+
+UPDATE observations set config = config || '{"sdr": "RTL-SDRv3"}' WHERE station_id = 1 and aos<'2021-06-09'::date;
```
Use "receiver" instead of "sdr". Some stations can use a dedicated devices to receive the radio signal.

tomaszmrugalski replied:
Renamed as suggested.

fivitti commented on server/db/svarog-11.psql:
```diff
@@ -0,0 +1,34 @@
+
+-- The last METEOR observation recorded is from 2021-03-26. That's because the newly installed LNA + filter
+-- (Nooelec NOAA) has a filter that is wide enough for NOAA, but too narrow for Meteor.
+UPDATE observations SET config = config || '{"lna":"Nooelec NOAA", "filter":"Nooelec NOAA"}'
```
Do we need the "lna" and "filter" both? It looks like duplicated data.

tomaszmrugalski replied:
Right now we have one device that has two features integrated, but I also have separate filter and separate LNA. Those can be different.

fivitti replied:
Ok, but the configuration isn't a fixed list, and you can omit the fields that aren't necessary. We use only the LNA now then we shouldn't add the filter field. 
If another station will have only a filter, it omits the LNA.

fivitti commented on server/templates/obslist.html:
```diff
@@ -68,6 +68,8 @@ <h3>Observations</h3>
         <th class="text-center">Rating</th>
         <th>Station</th>
+        <th>Frequency</th>
+        <th>Protocol</th>
```
I think that the protocol shouldn't be visible on the main list.

- The protocol Is little related to the frequency because the satellites use one protocol per frequency.
- The protocol may be unknown, but the frequency always is known.
- User may be unsure what protocol is used. For example, when the embedded metadata is received.

tomaszmrugalski replied:
I can easily hide it, but I was thinking that it's useful to compare APT to LRPT for example. People can draw conclusions from that. I was thinking that long term we would expose all of them, so you could filter and compare various scenarios, APT vs LRPT, reception with and without LNA etc. Also, having the protocol clickable (link to wikipedia) would have some educational value. Finally, the third argument is about emphasizing new protocols, which I very much hope we'll see soon.

So, does that sound convincing? If not, I'll remove it.

fivitti replied:
Ok, it makes sense. But we should be careful that the table isn't too wide.

fivitti commented on server/templates/obslist.html:
```diff
@@ -103,6 +103,17 @@ <h3>Observations</h3>
             </td>
+            <td>
+                {% if obs.config.freq is defined %}
+                {{ "%.3f MHz"|format(obs.config.freq) }}
```
Could you document that we expect the frequency in megahertz? Maybe, we should use just the hertz and use a template filter to convert the unit? I think that it will be the simplest and best solution.

tomaszmrugalski replied:
The docs now say it's in MHz. If you want, I can change it to Hz.

tomaszmrugalski replied:
After second though (and looking at what we use in the station's code), I agree that using Hz is the way to go. The observations list looks the same as before, but the underlying data is in Hz. You can see that when looking at specific observation page.

tomaszmrugalski replied:
Since I did exactly what was requested, I suppose it's ok to resolve.

fivitti commented on server/templates/station.html:
```diff
@@ -26,7 +26,14 @@ <h3>Ground Station  {{ station.name }}</h3>
+        <b>Current configuration:</b>
+        <ul>
+        {% for c in station.config %}
+        {% if station.config[c] %}
```
Use the `is defined` statement. Now, the value that equals `0` or `""` will not be visible.

tomaszmrugalski replied:
Added as suggested.

fivitti commented on server/tests/test_repository_postgres.py:
```diff
@@ -50,6 +50,10 @@ def test_db_version(self, repository):
             base, _ = os.path.splitext(filename)
+
+            # Skip any files that do not conform to the upgrade script naming convention
+            if base[:7] != "svarog-":
```
Use `.startswith` method of string.

tomaszmrugalski replied:
Updated as suggested.

tomaszmrugalski commented:
There's a handful of unrelated changes that were discovered or needed for other reasons:
- added setup and teardown functions to repository unit-tests. Don't do anything yet, but will be needed for extended tests
- corrected whitespace and some descriptions in pagination.py
- fixed exception on stations list if station didn't have registered field set
- updated schema, so registered field defaults to now() if not explicitly specified

I hope those are small enough. If you feel any of them should go to a separate PR, let me know.

tomaszmrugalski commented:
This is a request to do a preliminary review. I'm specifically interested if the ability to pass meta-data (called 'config', but perhaps calling it metadata would be more understandable?) is reasonable. I understand that there are multiple tests and comments missing. The feedback I'm looking for is for the concept.

There's a lot of changes:
- if run as stand-alone script, prints details on stdout
- handles various problems better, including printing response in case of failure
- MIME type is now generic, not limited to png
- one test added for submitobs (many still missing)
- support for png, jpg, gif, txt, csv, json files
- support for uploading data types that are NOT images (so thumbnail can't be generated)
- server logs more data during observation reception
- (unrelated) server no longer memory leaks when generating pass plots
- refactored ALLOWED_FILE_TYPES (because multiple extensions could lead to text/plain type)
- removed unused notes field
- added config field (technically it's a string, but should contain JSON. This is checked in the submitobs.py)

The bottom line is that you can now run the following command:
```python3 submitobs.py image1.png,image6.json "NOAA 18" "2021-11-27 21:35" "2021-11-27 21:37" "2021-11-27 21:39" '{"antenna":"corner", "receiver":"RTL-SDRv3", "frequency": 137100000 }' 0.3```

and it will upload multiple files (and possibly logs), will put the metadata into the observation and will generally work.

fivitti commented on server/app/controllers/receive.py:
```diff
@@ -42,17 +42,28 @@ class RequestArguments(TypedDict):
 get_extension = lambda p: os.path.splitext(p)[1]
-is_allowed_file = lambda f: f.mimetype in ALLOWED_FILE_TYPES and \
-                            ALLOWED_FILE_TYPES[f.mimetype] == get_extension(f.filename)
+is_allowed_file = lambda f: get_extension(f.filename) in ALLOWED_FILE_TYPES and \
```
Use: `ALLOWED_FILE_TYPES.get(get_extension(f.filename)) == f.mimetype)`

tomaszmrugalski replied:
Updated as suggested.

fivitti commented on server/app/controllers/receive.py:
```diff
@@ -104,19 +116,21 @@ def receive(station_id: str, args: RequestArguments):
-    thumb_source_filename, _ = thumbnail_source_entry
-    thumb_filename = "thumb-%s-%s" % (str(uuid.uuid4()), thumb_source_filename)
+        app.logger.info(f"No suitable images for thumbnail, using the default image {DEFAULT_THUMBNAIL}")
+        thumb_filename = DEFAULT_THUMBNAIL
```
Don't use a dedicated constant, just a `None` value.

tomaszmrugalski replied:
Removed DEFAULT_THUMBNAIL, using None.

fivitti commented on server/app/controllers/receive.py:
```diff
@@ -173,10 +187,12 @@ def receive(station_id: str, args: RequestArguments):
-    thumb_source_path = os.path.join(root, thumb_source_filename)
-    thumb_path = os.path.join(root, "thumbs", thumb_filename)
-    make_thumbnail(thumb_source_path, thumb_path)
+    if (thumb_filename != DEFAULT_THUMBNAIL):
```
Remove the brackets.

tomaszmrugalski replied:
Removed.

fivitti commented on server/db/svarog-11.psql:
```diff
@@ -0,0 +1,38 @@
+UPDATE observations SET config = config || '{"antenna": "WiMo TA-1", "antenna-type": "crossed dipole" }' WHERE station_id = 1;
+
+-- Add frequencies. We haven't done any Doppler magic, so everything was received on base frequencies.
+UPDATE observations SET config = config || '{"frequency": 137.62e6 }' WHERE sat_id = 25338 AND station_id = 1; -- NOAA 15
```
Frequency is in Hz or MHz, finally?

tomaszmrugalski replied:
Hz. I've updated the code as you suggested.

fivitti commented on server/templates/obslist.html:
```diff
@@ -103,6 +103,17 @@ <h3>Observations</h3>
             </td>
+            <td>
+                {% if obs.config.frequency is defined %}
+                {{ "%.3f MHz"|format(obs.config.frequency/1000000) }}
```
We will need a special frequency helper here (not necessarily now). The frequencies below 1MHz should be formatted in Hz, above 1GHz in GHz.

tomaszmrugalski replied:
Frequencies below 1MHz are useless for satellites, but I get your point. I'm not enthusiastic: mixing MHz and GHz may mess up sorting, but I can see how it could improve readability.

fivitti commented on server/tests/test_repository.py:
```diff
@@ -5,6 +5,14 @@
 
 class RepositoryTests(unittest.TestCase):
 
+    def setUp(self):
```
It should be added only if necessary.

fivitti commented on doc/submit_obs.md:
```diff
@@ -0,0 +1,21 @@
+## Configuration
+
+Configuration should be submitted as JSON. There's a number of parameters that
+are (almost) mandatory:
```
Maybe we should print the warnings when the "almost" mandatory field is missing?

tomaszmrugalski replied:
Added as suggested.

fivitti commented on station/receiver.py:
```diff
@@ -89,9 +89,10 @@ def cmd():
             rating = get_rating_for_product(product[1], satellite.get("rate"))
             logging.info("Product %s got rating %s" % (product[1], rating))
             # TODO: Submit ALL products and logs
+            files = [ product[1] ]
```
Maybe, we should send all products?

tomaszmrugalski replied:
That's a good and a bit risky idea. Do you want to upload 150MB+ wav files? So maybe not all, but most. In general, I totally agree it's a great idea. Created #66 for this. It's way too large for this ticket.

fivitti commented on station/submitobs.py:
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
+
+    txt = filename.lower()
+    offset = txt.rfind('.')
+    ext = txt[offset+1:] if offset != -1 else ""
```
Never ever Use `os.path.splitext` function.

tomaszmrugalski replied:
Done.

fivitti commented on station/submitobs.py:
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
+    txt = filename.lower()
+    offset = txt.rfind('.')
+    ext = txt[offset+1:] if offset != -1 else ""
+    return known_types[ext] if ext in known_types else "application/octet-stream"
```
Better: `known_types.get(ext, "application/octet-stream")`

tomaszmrugalski replied:
Done.

fivitti commented on station/submitobs.py:
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
     body: Dict[str, Any] = {
-        "file": file_obj
     }
+    for f in data.image_path:
```
`image_path` is a singular verb. But it's a list of paths now. Also, these paths mustn't be image paths.
In this context (reading a file), the `f` is confusing. The `f` means a file descriptor. It would help if you used a `path` verb.

tomaszmrugalski replied:
renamed as suggested.

fivitti commented on station/submitobs.py:
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
+
+        # If there's only one file, it will use "file" key. The second file will be "file1",
+        # third "file2" etc.
+        fkey = "file" if not cnt else f"file{cnt}"
```
`cnt == 0`

tomaszmrugalski replied:
Updated as suggested.

fivitti commented on station/submitobs.py:
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
+        fkey = "file" if not cnt else f"file{cnt}"
+
+        file_obj = open(f, 'rb')
+        body[fkey] = file_obj
```
If you named the above variable as `file_obj`, then, maybe, it should be `file_key`?

tomaszmrugalski replied:
Renamed.

fivitti commented on station/submitobs.py:
```diff
@@ -56,29 +62,81 @@ def get_tle(sat_name: str, date: datetime.datetime) -> Optional[List[str]]:
+        file_obj = open(f, 'rb')
+        body[fkey] = file_obj
+
+        files[fkey] = (filename, file_obj, get_mime_type(filename), {})
```
What is this empty dictionary at the end?

tomaszmrugalski replied:
I have no idea. This is what was there on master, before I touched it:

```
    files = {
        "file": (
            filename,
            file_obj,
            "image/png",
            {}
        )
    }
```

fivitti commented on station/submitobs.py:
```diff
@@ -88,41 +146,45 @@ def submit_observation(data: SubmitRequestData):
+    logging.debug(f"Meta-data = {data.config}")
 
-    logging.info("Submitting observation, url=%s file=%s" % (url, filename))
+    logging.info(f"Submitting observation, url={url}, file(s)={data.image_path}")
```
It looks verbose. Use the debug level.

tomaszmrugalski replied:
This is very useful piece of information and usually there's only one file. Are you sure about debug?

fivitti commented on station/submitobs.py:
```diff
@@ -88,41 +146,45 @@ def submit_observation(data: SubmitRequestData):
+    return None
 
 if __name__ == '__main__':
     if len(sys.argv) < 4:
```
It has grown a lot. We need to refactor it using the `argparse` module. But maybe in a separate issue?

tomaszmrugalski replied:
There's #57 created for that already.

tomaszmrugalski commented:
Ok, last batch of changes is pushed. It adds Metadata class that maintains the metadata on station side and actually uses it for observation submissions.

tomaszmrugalski commented:
@fivitti is busy with some other task that needs to be completed in 2021. I'll merge this now, but will keep the threads open, so it will be reasonably easy to get back to the discussions later, if need be.
